llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp

```diff
@@ -0,0 +1,161 @@
+//===-- AIEEliminateDuplicatePHI.cpp - Identify and remove duplicate PHI --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+// Identify and remove duplicate PHI nodes in a basic block.
+// Duplicate PHI nodes are converted in to copy instructions.
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseSubtarget.h"
+#include "llvm/CodeGen/GlobalISel/CSEInfo.h"
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+#include <set>
+#include <unordered_map>
+
+#define DEBUG_TYPE "aie-duplicate-PHI-elimination"
+
+using namespace llvm;
+
+static const char AIE_ELIMINATE_DUPLICATE_PHI[] =
+    "AIE Eliminate Duplicate PHI Pass";
+
+namespace {
+
+class AIEEliminateDuplicatePHI : public MachineFunctionPass {
```

Comment:
Nit: We could maybe have that as part of the addrspace retyping pass, because it's really there that the duplicates are introduced. Not sure what's better. I'm fine to keep it as a separate pass.

Reply:
After the address space retype pass, PHIs are not in state where they can be classified duplicate. After our pre-legalizer combiner passes they come into state where they are easy to classify as duplicate.

I have tried to add early checks to and use unorded_map to minimize the impact of the pass.